lightning/src/ln/channelmanager.rs

```diff
@@ -71,6 +71,263 @@ use core::time::Duration;
+				unknown => Err(unknown),
+			}
+		}
+	}
+
+	pub(super) fn create<Signer: Sign, K: Deref>(keys: &ExpandedKey, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, keys_manager: &K, highest_seen_timestamp: u64) -> Result<(PaymentHash, PaymentSecret), ()>
+		where K::Target: KeysInterface<Signer = Signer>
+	{
+		let metadata_bytes = construct_metadata_bytes(min_value_msat, Method::LdkPaymentHash, invoice_expiry_delta_secs, highest_seen_timestamp)?;
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		let rand_bytes = keys_manager.get_secure_random_bytes();
+		iv_bytes.copy_from_slice(&rand_bytes[..IV_LEN]);
+
+		let mut hmac = HmacEngine::<Sha256>::new(&keys.ldk_pmt_hash_key);
+		hmac.input(&iv_bytes);
+		hmac.input(&metadata_bytes);
+		let payment_preimage_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let ldk_pmt_hash = PaymentHash(Sha256::hash(&payment_preimage_bytes).into_inner());
+		let payment_secret = construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key);
+		Ok((ldk_pmt_hash, payment_secret))
+	}
+
+	pub(super) fn create_from_hash(keys: &ExpandedKey, min_value_msat: Option<u64>, payment_hash: PaymentHash, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<PaymentSecret, ()> {
+		let metadata_bytes = construct_metadata_bytes(min_value_msat, Method::UserPaymentHash, invoice_expiry_delta_secs, highest_seen_timestamp)?;
+
+		let mut hmac = HmacEngine::<Sha256>::new(&keys.user_pmt_hash_key);
+		hmac.input(&metadata_bytes);
+		hmac.input(&payment_hash.0);
+		let hmac_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		iv_bytes.copy_from_slice(&hmac_bytes[..IV_LEN]);
+
+		Ok(construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key))
+	}
+
+	fn construct_metadata_bytes(min_value_msat: Option<u64>, payment_type: Method, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<[u8; METADATA_LEN], ()> {
+		if min_value_msat.is_some() && min_value_msat.unwrap() > MAX_VALUE_MSAT {
+			return Err(());
+		}
+
+		let mut min_amt_msat_bytes: [u8; AMT_MSAT_LEN] = match min_value_msat {
+			Some(amt) => amt.to_be_bytes(),
+			None => [0; AMT_MSAT_LEN],
+		};
+		min_amt_msat_bytes[0] |= (payment_type as u8) << METHOD_TYPE_OFFSET;
+
+		// We assume that highest_seen_timestamp is pretty close to the current time - it's updated when
+		// we receive a new block with the maximum time we've seen in a header. It should never be more
+		// than two hours in the future.  Thus, we add two hours here as a buffer to ensure we
+		// absolutely never fail a payment too early.
+		// Note that we assume that received blocks have reasonably up-to-date timestamps.
+		let expiry_bytes = (highest_seen_timestamp + invoice_expiry_delta_secs as u64 + 7200).to_be_bytes();
+
+		let mut metadata_bytes: [u8; METADATA_LEN] = [0; METADATA_LEN];
+		{
+			let (min_amt_msat_slice, expiry_slice) = metadata_bytes.split_at_mut(AMT_MSAT_LEN);
```

Comment:
nit: I find `split_at_mut` rather hard to read. Instead, you can just change the lines below to `metadata_bytes[..AMT_MSAT_LEN].copy_from_slice(..)` and `[AMT_MSAT_LEN..]`.